### DISCUSSION TITLE
Is rendering to many different canvas elements supposed to be slow?

### DISCUSSION THREAD

**Discussion by ziriax** [gpuweb/gpuweb#1705] [Q&A] [answered]
I'm trying out WebGPU, I love the spec so far! It is really nice to have a new modern developer-friendly 3D API.

I liked OpenGL a lot when programming on Silicon Graphics machines, and then on Intergraph PCs and Wildcat GPUs back in the 90's, but then switched to Direct3D. 

My experiment is to see if WebGPU is suitable for rendering to many different canvas elements. 

After all, the web is about documents, and a page can have many 3D elements on it.

I'm rendering just a single triangle to many tiny 64x64 canvas elements, updating a single uniform variable.

With WebGL I already found it this is way too slow (I render to a single canvas and then use `drawImage`). 

With WebGPU, on my GTX1080 card, I can have about 18 canvasses before the framerate drops below 60, and wit 64 canvas elements I get 30 FPS. I'm using a new swapchain per canvas.

Is this performance to be expected?

What would be the fastest way to render to many canvas elements on the same page? Render to a single offscreen texture, and copy?

It would be nice if the DOM had a "canvas slice" element, like in the old days, before it got deprecated in webkit...

**Comment by ziriax** [comment]
Actually, I was wrong, after a bit of refactoring the WebGL version is faster. 

Also, it seems that I can't just use `context2D.drawImage(canvas3D, 0, 0)` after dispatching the commands in the WebGPU version, even not after `onSubmittedWorkDone`; inserting `copyTextureToTexture` did work but was very slow.

I'll try to put the code online, that will make it easier to discuss about this.

**Comment by Kangz** [comment]
> With WebGPU, on my GTX1080 card, I can have about 18 canvasses before the framerate drops below 60, and wit 64 canvas elements I get 30 FPS. I'm using a new swapchain per canvas.
>
> Is this performance to be expected?

This is a bit unexpected, WebGPU should be able to render to many canvases with little performance penalty. I assume you're using Chrome since Firefox doesn't yet implement `onSubmittedWorkDone`.  I think what you might be hitting is that we don't yet recycle the textures used by the swapchains, so each new frame will causes a fairly expensive allocation to occur, while this is optimized in WebGL already. Would you be able to share a link to the demo so we can run a trace and see what's expensive?

> What would be the fastest way to render to many canvas elements on the same page? Render to a single offscreen texture, and copy?

Having a swapchain per canvas should work well. Although you probably want to stop rendering to a canvas when it gets outside of the viewport (using the [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) API)

**Comment by ziriax** [comment]
Thanks for your feedback!

Yes indeed I'm using Chrome on Windows 10. 

I will share my code asap. I already used a WeakMap to avoid creating a new texture view for each swap chain every frame, but if the texture pointer is unstable, this won't help. 

PS: I am already using intersection observer on my site, thanks for the hint, I am just benchmarking here.

**Reply by ziriax** [reply]
I shared my code here: https://stackblitz.com/edit/typescript-jj8vpx?file=index.ts

It also runs very slow on my RTX2070.

And you are right, the swap texture seems to be recreated every frame (I checked this with a WeakMap)

PS: Might be handy for others to set their first steps, I added some error validation, and an error log

**Reply by Kangz** [reply]
The swapchain is meant to always give you a new JS object, however the underlying storage should be recycled between frames, but Chrome doesn't implement that yet.

**Reply by Kangz** [reply]
Thanks for the link, I'll take a look to see if it is indeed this.

**Reply by ziriax** [reply]
Thanks!

PS: I'm cloning Dawn; this is a good exercise for me to try to debug this.

**Reply by Kangz** [reply]
This part of the code will be in Chromium: here https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/graphics/gpu/webgpu_swap_buffer_provider.cc

**Comment by kvark** [comment]
I'm curious about this use case. Could you elaborate why you want to have many canvases?

**Reply by Kangz** [reply]
A site like this one maybe? https://ciechanow.ski/internal-combustion-engine/

**Reply by kvark** [reply]
There is never more than 2 canvases at a time on screen there, at least within the 100% zoom.

**Reply by ziriax** [reply]
We are building an educational reactive visual programming website and web-game, funded by the Flemish Government (VAF Gamefund), where each output of each reactive function is rendered to a canvas, and each output is potentially animating. So we end up with a lot of canvas elements 😉

So far this works reasonably well for 2D graphics. We have been working on this for the last 9 years, albeit part-time. We are investigating if it is possible to support 3D, in the same way.

It would be better to render everything to a single canvas, just changing the viewport or scissors, and then have some DOM element that renders a slice from the big canvas. Then you have just two render-targets instead of one render target per canvas. Unfortunately as far as I can tell, this feature doesn't exist anymore; it used to be experimentally supported, but I can't remember the name. E.g. in Microsoft's WPF, one could use a `VisualBrush` for this.

**Comment by ziriax** [comment]
I tried this again today to see if WebGPU is ready for our webapp, and rendering a simple rotating triangle to many canvasses on the page is still very slow.

Too bad, I really would have loved to add 3D support to my education webapp, but it heavily depends on many canvas objects.

**Comment by ziriax** [comment]
Oh! I found out that rendering to a single large MSAA texture, using `setScissorRect` and `setViewport` to draw each cell, and then after all the rendering using `copyTextureToTexture` to draw to the canvas textures makes it faster; on my laptop with A3000 GPU I get about 60 FPS for 100 canvas objects. It drops to 30 FPS for 200 canvasses. It still feels slow, I don't see a reason for the slowdown, but certainly good enough to finally get started.

I shared my code here: https://github.com/ziriax/webgpu-multi-canvas-test

All tips are welcome to improve this 😉

**Comment by greggman** [comment, marked as answer]
I'd try only rendering to canvases that are visible.

I think this works


      const rect = canvas.getBoundingClientRect();
      if (rect.bottom < 0 || rect.top  > window.innerHeight ||
          rect.right  < 0 || rect.left > window.innerWidth) {
        return;  // it's off screen
      }


https://jsgist.org/?src=81c0edae17229d8dff5467f5385802fe

That's 400 canvases. It's rendering at 60fps for me but of course that's my CPU/GPU combo

Of course all the usual WebGPU (gpu in general) things apply. If a resource is in use and you try to update it the system may have to stall. In the example linked there are only 2 uniform buffers that are reused for every canvas. That means when rendering the 2nd canvas, before the system can update those 2 buffers it has to pause and wait for them to be unused (*). It would be better to create 2 uniform buffers PER cube. Then there would be no waiting. I didn't add that optimization because, at least for me, the sample already ran fast enough. I'd need to test to see how much having uniform buffers per cube would help.

**Comment by greggman** [comment]
I thought I add that rather then `getBoundingClientRect`, you can use [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) to track which canvases are visible.

```
  const visibleCanvasSet = new Set();
  const observer = new IntersectionObserver((entries) => {
    for (const {target, isIntersecting } of entries) {
      if (isIntersecting) {
        visibleCanvasSet.add(target);
      } else {
        visibleCanvasSet.delete(target);
      }
    }
  });
```

then observe each canvas

```
for each canvas
     observer.observe(canvas);
```

Then at render time just render the visible canvases

```
    visibleCanvasSet.forEach((canvas) => {
       ... render to canvas ...
    }
```

Example: https://jsgist.org/?src=ff40db4cc836b2f4d1e041fde326dd72

**Reply by ziriax** [reply]
Thanks for the suggestion. I am already using the Intersection observer API, I just have many small visible canvasses on the screen.

**Reply by greggman** [reply]
If it fits you needs you can use the solution outlined at the bottom of this article

https://webglfundamentals.org/webgl/lessons/webgl-multiple-views.html

It's similar to your scissor/viewport but avoids the copies by using a single canvas. 

Even if there are optimizations to be made for multiple canvases in WebGPU, if you have 200 canvases on the screen that's 200 things that have to be done (200 textures to render to, 200 textures to synchronize with the compositor, 200 textures the draw into the page), where as with the technique in the article above those 200 things become 1 thing (one texture to render to, one texture to synchronize, one texture to draw into the page) so it should always be faster.

Here's 2000 "canvases"

https://jsgist.org/?src=241fb67d7615fa723516303d4cb26938

**Reply by ziriax** [reply]
That is a nice technique, but in my case, I have many different canvas elements in separate scrolling containers.

It's actually the first thing I tried.

When following that path, I would prefer to get rid of all HTML and CSS and switch to rendering everything with WebGPU, like the amazing https://usegpu.live. I never liked HTML/CSS anyway, I just use it because it is "the defacto standard".

Anyway, I don't have much time to play with WebGPU anymore since I joined NVIDIA to work on Omnigraph/Omniverse.
